src/serde/object_cache.rs

```diff
@@ -0,0 +1,268 @@
+/// `ObjectCache` provides a way to calculate and cache values for each node
+/// in a clvm object tree. It can be used to calculate the sha256 tree hash
+/// for an object and save the hash for all the child objects for building
+/// usage tables, for example.
+///
+/// It also allows a function that's defined recursively on a clvm tree to
+/// have a non-recursive implementation (as it keeps a stack of uncached
+/// objects locally).
+use crate::allocator::{Allocator, NodePtr, SExp};
+type CachedFunction<T> = fn(&mut ObjectCache<T>, &Allocator, NodePtr) -> Option<T>;
+use super::bytes32::{hash_blobs, Bytes32};
+
+pub struct ObjectCache<'a, T> {
+    cache: Vec<Option<T>>,
+    allocator: &'a Allocator,
+
+    /// The function `f` is expected to calculate its T value recursively based
+    /// on the T values for the left and right child for a pair. For an atom, the
+    /// function f must calculate the T value directly.
+    ///
+    /// If a pair is passed and one of the children does not have its T value cached
+    /// in `ObjectCache` yet, return `None` and f will be called with each child in turn.
+    /// Don't recurse in f; that's the point of this structure.
+    f: CachedFunction<T>,
+}
+
+/// turn a `NodePtr` into a `usize`. Positive values become even indices
+/// and negative values become odd indices.
+
+fn node_to_index(node: &NodePtr) -> usize {
+    let node = *node;
+    if node < 0 {
+        (-node - node - 1) as usize
+    } else {
+        (node + node) as usize
+    }
+}
+
+impl<'a, T: Clone> ObjectCache<'a, T> {
+    pub fn new(allocator: &'a Allocator, f: CachedFunction<T>) -> Self {
+        let cache = vec![];
+        Self {
+            cache,
+            allocator,
+            f,
+        }
+    }
+
+    /// return the function value for this node, either from cache
+    /// or by calculating it
+    pub fn get_or_calculate(&mut self, node: &NodePtr) -> Option<&T> {
+        self.calculate(node);
+        self.get_from_cache(node)
+    }
+
+    /// return the cached value for this node, or `None`
+    fn get_from_cache(&self, node: &NodePtr) -> Option<&T> {
+        let index = node_to_index(node);
+        if index < self.cache.len() {
+            self.cache[index].as_ref()
+        } else {
+            None
+        }
+    }
+
+    /// set the cached value for a node
+    fn set(&mut self, node: &NodePtr, v: T) {
+        let index = node_to_index(node);
+        if index >= self.cache.len() {
+            self.cache.resize(index + 1, None);
```

Comment:
For consensus, naw, because we're charging reasonable costs for the decompression, and the memory usage is pretty trivial compared to DOS attack.

In general, I think memory usage for legitimate actions is going to be dwarfed by DOS attacks. In other words, if we're doing a real, useful chia contract, memory and CPU usage will be quite reasonable. But DOS attacks can use as much as they want and we can't really do much about it (beside soft forking in further limits in future).